Add "Ignore Date" and "Committer Date Is Author Date" options to rebase dialog

gtisdelle opened the pull request:
<!-- Please read CONTRIBUTING.md before submitting a pull request -->

Fixes #7708 


## Proposed changes

- [x]  Add "Ignore Date" option to rebase dialog

- [x]  Add tooltip for "Ignore Date"

- [x] Add "Committer Date is Author Date" option to rebase dialog (Suggested by @pmiossec [here](https://github.com/gitextensions/gitextensions/issues/7708#issuecomment-652452736))

- [x] Add tooltip for "Committer Date is Author Date"


## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

![before](https://user-images.githubusercontent.com/54757231/86497064-14115380-bd4e-11ea-8354-5ac9e1afc4f6.jpg)


### After

![after](https://user-images.githubusercontent.com/54757231/87229707-7be02380-c378-11ea-8aca-1134e8e677be.jpg)


## Test methodology <!-- How did you ensure quality? -->

- Regression testing using the test suites provided
- Added test cases to test suite
- System testing on the rebase dialog


## Test environment(s) <!-- Remove any that don't apply -->

- Git 2.27.0.windows.1
- Microsoft Windows NT 10.0.18363.0
- .NET Framework 4.8.4180.0
- DPI 192dpi (200% scaling)

<!-- Mention language, UI scaling, or anything else that might be relevant -->


----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

mstv commented:
Run `build.cmd -loc`, please. Then commit the updated `English.xlf`.

mstv commented on GitCommands/Git/GitCommandHelpers.cs:
```diff
@@ -436,10 +436,11 @@ public static ArgumentString RebaseCmd(string branch, bool interactive, bool pre
-                { interactive && !autosquash, "--no-autosquash" },
-                { preserveMerges, GitVersion.Current.SupportRebaseMerges ? "--rebase-merges" : "--preserve-merges" },
+                { ignoreDate, "--ignore-date" },
+                { interactive && !ignoreDate, "-i" },
+                { interactive && autosquash && !ignoreDate, "--autosquash" },
+                { interactive && !autosquash && !ignoreDate, "--no-autosquash" },
+                { preserveMerges && !ignoreDate, GitVersion.Current.SupportRebaseMerges ? "--rebase-merges" : "--preserve-merges" },
```
nitpick: I'd put `!ignoredate' in front in order the emphasize the incompatibility. Thoughts?

RussKie replied:
I don't have a strong preference.

Maybe it worth structuring differently, e.g.:
```cs
            var builder = new GitArgumentBuilder("rebase");
            if (ignoreDate)
            {
                builder.Add("--ignore-date");
            }
            else
            {
                builder.AddRange({ ... });
            }

            builder.AddRange({ /* common */ });
```

mstv replied:
Yes, this will be clearer.

gtisdelle replied:
I agree. As much as I appreciate the initializer, I think it's fairly unclear to use with this kind of logic. I think going back to a standard selection statement is the way to go, especially since I am also planning on adding another checkbox with the same types of incompatibilities.

mstv commented on GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:
```diff
@@ -2593,7 +2593,7 @@ private void LaunchRebase(string command)
             string rebaseCmd = GitCommandHelpers.RebaseCmd(
                 LatestSelectedRevision.FirstParentId?.ToString(),
-                interactive: true, preserveMerges: false, autosquash: false, autoStash: true);
+                interactive: true, preserveMerges: false, autosquash: false, autoStash: true, ignoreDate: true);
```
This is too intrusive IMO because `interactive` and `ignoreDate` conflict with each other.

mstv replied:
I.e. it overrides the previous behavior because `ignoreDate` takes precedence.
```suggestion
                interactive: true, preserveMerges: false, autosquash: false, autoStash: true, ignoreDate: false);
```

pmiossec commented:
2 quick things:

1. It seems there are too many changes in `FormRebase.Designer.cs`. Please launch visual studio in no scale (start VS with ` /noscale` parameter) mode before opening the designer and commit the changes.

>Add "Committer Date is Author Date" option to rebase dialog (Suggested by @pmiossec here)

2. You announce that but I don't see it in the changes. Is it because that's a work in progress that it is not in the PR and you will add it later? Or there is a commit missing?

gtisdelle commented:
> 
> 
> 2 quick things:
> 
>     1. It seems there are too many changes in `FormRebase.Designer.cs`. Please launch visual studio in no scale (start VS with ` /noscale` parameter) mode before opening the designer and commit the changes.
> 
> 
> > Add "Committer Date is Author Date" option to rebase dialog (Suggested by @pmiossec here)
> 
>     2. You announce that but I don't see it in the changes. Is it because that's a work in progress that it is not in the PR and you will add it later? Or there is a commit missing?

1. Maybe some of those changes have to do with the fact that I changed the default size of the form? I made sure to restart VS in non-scaling mode before opening the designer. I also just opened it with that parameter, and there was nothing to commit.

2. You're right that was a bit confusing. I updated the PR to better show that adding "Commit Date is Author Date" is a future task.

gtisdelle commented:
> 
> 
> Run `build.cmd -loc`, please. Then commit the updated `English.xlf`.

:heavy_check_mark: Done

gtisdelle commented:
I will checkout the master branch's `FormRebase.Designer.cs` and redo the changes to that file in the designer while making sure to launch Visual Studio with the /noscale parameter within the next few days. Hopefully that will solve it. Thanks for that advice,  @pmiossec and @RussKie!

RussKie commented on GitCommands/Git/GitCommandHelpers.cs:
```diff
@@ -427,24 +427,45 @@ public static ArgumentString StopBisectCmd()
+                builder.Add("--autostash");
+            }
+
+            builder.AddRange(new List<string>() { from.QuoteNE(), branch.Quote() });
```
This will cause unnecessary allocations.
For the common path I'd keep the original syntax.

gtisdelle replied:
Do you mean the List declaration on line 461?

gtisdelle replied:
Looking at it again, I was able to use the (boolean, string) extension method to simplify the common path. I also eliminated the useless List initialization.

mstv commented on GitUI/CommandsDialogs/FormRebase.Designer.cs:
```diff
@@ -264,13 +270,26 @@ private void InitializeComponent()
+            this.chkIgnoreDate.Name = "chkIgnoreDate";
+            this.chkIgnoreDate.Size = new System.Drawing.Size(82, 17);
+            this.chkIgnoreDate.TabIndex = 9;
+            this.chkIgnoreDate.Text = "Ignore Date";
```
```suggestion
            this.chkIgnoreDate.Text = "Ignore date";
```

mstv commented on GitUI/CommandsDialogs/FormRebase.Designer.cs:
```diff
@@ -542,30 +571,34 @@ private void InitializeComponent()
+            this.chkCommitterDateIsAuthorDate.Name = "chkCommitterDateIsAuthorDate";
+            this.chkCommitterDateIsAuthorDate.Size = new System.Drawing.Size(169, 17);
+            this.chkCommitterDateIsAuthorDate.TabIndex = 10;
+            this.chkCommitterDateIsAuthorDate.Text = "Committer Date Is Author Date";
```
```suggestion
            this.chkCommitterDateIsAuthorDate.Text = "Committer date is author date";
```

RussKie commented on GitUI/CommandsDialogs/FormRebase.Designer.cs:
```diff
@@ -264,13 +270,26 @@ private void InitializeComponent()
+            this.chkIgnoreDate.Text = "Ignore date";
+            this.toolTip1.SetToolTip(this.chkIgnoreDate, "Changes the author dates of the commits in this branch\r\ninto the commit dates.");
+            this.chkIgnoreDate.UseVisualStyleBackColor = true;
+            this.chkIgnoreDate.Click += new System.EventHandler(this.DateClick);
```
```suggestion
            this.chkIgnoreDate.Click += new System.EventHandler(this.chkIgnoreDate_Click);
```

RussKie commented on GitUI/CommandsDialogs/FormRebase.Designer.cs:
```diff
@@ -505,6 +524,16 @@ private void InitializeComponent()
+            this.Commit.TabIndex = 23;
+            this.Commit.Text = "Commit...";
+            this.Commit.UseVisualStyleBackColor = true;
+            this.Commit.Click += new System.EventHandler(this.CommitButtonClick);
```
```suggestion
            this.Commit.Click += new System.EventHandler(this.Commit_Click);
```

RussKie commented on GitUI/CommandsDialogs/FormRebase.Designer.cs:
```diff
@@ -542,30 +571,34 @@ private void InitializeComponent()
+            this.toolTip1.SetToolTip(this.chkCommitterDateIsAuthorDate, "Changes the committer dates of the commits in this branch\r\ninto the author dates." +
+        "\r\n");
+            this.chkCommitterDateIsAuthorDate.UseVisualStyleBackColor = true;
+            this.chkCommitterDateIsAuthorDate.Click += new System.EventHandler(this.DateClick);
```
Please don't reuse the same handler, create a new one and call a shared method. Otherwise it makes it extremely difficult to reason about the code.
```suggestion
            this.chkCommitterDateIsAuthorDate.Click += new System.EventHandler(this.chkCommitterDateIsAuthorDate_Click);
```

RussKie commented on GitUI/CommandsDialogs/FormRebase.cs:
```diff
@@ -182,6 +182,15 @@ private void InteractiveRebaseClick(object sender, EventArgs e)
+            chkInteractive.Enabled = !chkIgnoreDate.Checked && !chkCommitterDateIsAuthorDate.Checked;
+            chkPreserveMerges.Enabled = !chkIgnoreDate.Checked && !chkCommitterDateIsAuthorDate.Checked;
+            chkAutosquash.Enabled = chkInteractive.Checked && !chkIgnoreDate.Checked && !chkCommitterDateIsAuthorDate.Checked;
+        }
```
I'd like to see a test for this functionality since it embodies the git command logic, which is defined elsewhere. 

Please follow the `TestAccessor` pattern (search the codebase for it), expose these UI elements through it and have tests something like:
```cs
[Test]
public void Check_ignore_date_should_toggle_non_applicable_settings()
{
    CheckBox chkIgnoreDate = form.GetTestAccessor().chkIgnoreDate;
    chkIgnoreDate.Checked.Should().BeFalse();

    chkIgnoreDate.Checked = true;

    chkIgnoreDate.Enabled.Should().BeFalse();
    // etc.

    chkIgnoreDate.Checked = false;

    chkIgnoreDate.Enabled.Should().BeTrue();
    // etc.
}
```

gtisdelle commented:
It looks like I will be creating the unit test for this class. That means that things like the Interactive Merge and Autosquash toggle are not tested currently. Should I test them too, or would that be outside the scope of this PR?

pmiossec commented:
Most of the contributors agrees to say that there are not enough tests.

So I'm not sure someone will refuse the proposition 😁

RussKie commented on GitUI/CommandsDialogs/FormRebase.Designer.cs:
```diff
@@ -264,13 +270,40 @@ private void InitializeComponent()
+            this.chkCommitterDateIsAuthorDate.TabIndex = 10;
+            this.chkCommitterDateIsAuthorDate.Text = "Committer date is author date";
+            this.toolTip1.SetToolTip(this.chkCommitterDateIsAuthorDate, "Changes the committer dates of the commits in this branch\r\ninto the author dates." +
+        "\r\n");
```
Is the last "\r\n" desirable?

RussKie commented on GitUI/CommandsDialogs/FormRebase.cs:
```diff
@@ -314,10 +333,32 @@ private void btnChooseFromRevision_Click(object sender, EventArgs e)
+            public CheckBox chkStash => _form.chkStash;
+            public CheckBox chkIgnoreDate => _form.chkIgnoreDate;
+            public CheckBox chkCommitterDateIsAuthorDate => _form.chkCommitterDateIsAuthorDate;
+            public void chkIgnoreDate_Click() => _form.chkIgnoreDate_Click(_form.chkIgnoreDate, null);
+            public void chkCommitterDateIsAuthorDate_Click() => _form.chkCommitterDateIsAuthorDate_Click(_form.chkCommitterDateIsAuthorDate, null);
+            public void InteractiveRebaseClick() => _form.InteractiveRebaseClick(_form.chkInteractive, null);
```
These are redundant. The handlers are invoked by the virtue of changing the `Checked` state.

gtisdelle replied:
I thought so at first too, but the `Click` event is not fired by changing the `Checked` property. I even experimented with it in my own project to see, and it seems like only the `CheckChanged` event is fired by changing the `Checked` property programmatically.

RussKie commented on IntegrationTests/UI.IntegrationTests/CommandsDialogs/FormRebaseTests.cs:
```diff
@@ -0,0 +1,216 @@
+                    var accessor = form.GetTestAccessor();
+
+                    accessor.chkIgnoreDate.Checked = true;
+                    accessor.chkIgnoreDate_Click();
```
As above, this call is redundant.

RussKie commented on GitUI/CommandsDialogs/FormRebase.Designer.cs:
```diff
@@ -264,13 +270,39 @@ private void InitializeComponent()
+            this.chkCommitterDateIsAuthorDate.Text = "Committer date is author date";
+            this.toolTip1.SetToolTip(this.chkCommitterDateIsAuthorDate, "Changes the committer dates of the commits in this branch\r\ninto the author dates.");
+            this.chkCommitterDateIsAuthorDate.UseVisualStyleBackColor = true;
+            this.chkCommitterDateIsAuthorDate.Click += new System.EventHandler(this.chkCommitterDateIsAuthorDate_Click);
```
Apologies, I totally missed it.
The correct event here is [`CheckedChanged`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.checkbox.checkedchanged?view=netframework-4.6.1)

gtisdelle replied:
Yeah it seems like all the check-boxes on the form should have been using `CheckedChanged` events in the first place. I'll change the one for `chkInteractive` along with changing the ones for `chkIgnoreDate` and `chkCommitterDateIsAuthorDate`. If nothing else, it will simplify the test suite.

RussKie commented:
Thank you

On Thu, Jul 16, 2020, 2:55 PM George Tisdelle <notifications@github.com>
wrote:

> *@gtisdelle* commented on this pull request.
> ------------------------------
>
> In GitUI/CommandsDialogs/FormRebase.Designer.cs
> <https://github.com/gitextensions/gitextensions/pull/8289#discussion_r455509633>
> :
>
> > +            this.toolTip1.SetToolTip(this.chkIgnoreDate, "Changes the author dates of the commits in this branch\r\ninto the commit dates.");
> +            this.chkIgnoreDate.UseVisualStyleBackColor = true;
> +            this.chkIgnoreDate.Click += new System.EventHandler(this.chkIgnoreDate_Click);
> +            //
> +            // chkCommitterDateIsAuthorDate
> +            //
> +            this.chkCommitterDateIsAuthorDate.Anchor = System.Windows.Forms.AnchorStyles.Left;
> +            this.chkCommitterDateIsAuthorDate.AutoSize = true;
> +            this.chkCommitterDateIsAuthorDate.Location = new System.Drawing.Point(493, 3);
> +            this.chkCommitterDateIsAuthorDate.Name = "chkCommitterDateIsAuthorDate";
> +            this.chkCommitterDateIsAuthorDate.Size = new System.Drawing.Size(163, 17);
> +            this.chkCommitterDateIsAuthorDate.TabIndex = 10;
> +            this.chkCommitterDateIsAuthorDate.Text = "Committer date is author date";
> +            this.toolTip1.SetToolTip(this.chkCommitterDateIsAuthorDate, "Changes the committer dates of the commits in this branch\r\ninto the author dates.");
> +            this.chkCommitterDateIsAuthorDate.UseVisualStyleBackColor = true;
> +            this.chkCommitterDateIsAuthorDate.Click += new System.EventHandler(this.chkCommitterDateIsAuthorDate_Click);
>
> Yeah it seems like all the check-boxes on the form should have been using
> CheckedChanged events in the first place. I'll change the one for
> chkInteractive along with changing the ones for chkIgnoreDate and
> chkCommitterDateIsAuthorDate. If nothing else, it will simplify the test
> suite.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/8289#discussion_r455509633>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXVZZAMQRKEBS6KMVGDR32B47ANCNFSM4OQCEJRA>
> .
>

RussKie commented:
Please squash and it is ready to go

gtisdelle commented:
All the changes have been squashed into one functional commit. Thank you for all your help!
